### ISSUE TITLE
Exception serialization fails when CELERY_RESULT_SERIALIZER != 'pickle'

### ISSUE THREAD

**Issue by dbrgn** [celery/celery#2409]
I'm on Celery 3.1.16. I'm using the following serialization settings:

    CELERY_ACCEPT_CONTENT = ['msgpack', 'json']
    CELERY_TASK_SERIALIZER = 'msgpack'
    CELERY_RESULT_SERIALIZER = 'msgpack'

As result backend I use Redis with `new_join=1` (which I enabled a few days ago).

Recently I started getting intermittent encoding errors, like the following two:

```python
Task worker.tasks.process_account with id 19f472cb-d739-4a97-9810-c72f70c9acc1 raised exception:
"UnicodeDecodeError('ascii', '\\x84\\xa6status\\xa7STARTED\\xa9traceback\\xc0\\xa6result\\x82\\xa8hostname\\xbfcelery@celery-0.ip-172-31-27-15\\xa3pid\\xcd\\x08\\x1b\\xa8children\\x90', 0, 1, 'ordinal not in range(128)')"
 
 
Task was called with args: [34, 3490427483, u'8f6557b09aeb49c2b8652cbf8adf300d'] kwargs: {}.
 
The contents of the full traceback was:
 
Traceback (most recent call last):
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/app/trace.py", line 235, in trace_task
    request=task_request,
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/base.py", line 256, in store_result
    request=request, **kwargs)
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/base.py", line 486, in _store_result
    self.set(self.get_key_for_task(task_id), self.encode(meta))
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/redis.py", line 160, in set
    return self.ensure(self._set, (key, value), **retry_policy)
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/redis.py", line 149, in ensure
    **retry_policy
  File "/var/venv/radar/local/lib/python2.7/site-packages/kombu/utils/__init__.py", line 243, in retry_over_time
    return fun(*args, **kwargs)
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/redis.py", line 169, in _set
    pipe.execute()
  File "/var/venv/radar/local/lib/python2.7/site-packages/redis/client.py", line 2578, in execute
    return execute(conn, stack, raise_on_error)
  File "/var/venv/radar/local/lib/python2.7/site-packages/redis/client.py", line 2465, in _execute_transaction
    self.annotate_exception(ex, i + 1, command[0])
  File "/var/venv/radar/local/lib/python2.7/site-packages/redis/client.py", line 2529, in annotate_exception
    cmd = unicode(' ').join(imap(unicode, command))
UnicodeDecodeError: 'ascii' codec can't decode byte 0x84 in position 0: ordinal not in range(128)
```

```python
Task worker.tasks.process_account with id ff4152f2-f9f0-4ba0-863f-677bfb2f0f61 raised exception:
'EncodeError(TypeError("can\'t serialize <ExceptionInfo: UnicodeDecodeError(\'ascii\', \'Article non trouv\\\\xc3\\\\xa9\', 17, 18, \'ordinal not in range(128)\')>",),)'


Task was called with args: [41, 8722504955, u'003eb1865b6f4e7e98af73bcafb749de'] kwargs: {}.

The contents of the full traceback was:

Traceback (most recent call last):
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/app/trace.py", line 296, in trace_task
    on_chord_part_return(task, state, R)
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/redis.py", line 209, in _new_chord_return
    .rpush(jkey, self.encode([1, tid, state, result]))          \
  File "/var/venv/radar/local/lib/python2.7/site-packages/celery/backends/base.py", line 182, in encode
    _, _, payload = dumps(data, serializer=self.serializer)
  File "/var/venv/radar/local/lib/python2.7/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/usr/lib/python2.7/contextlib.py", line 35, in __exit__
    self.gen.throw(type, value, traceback)
  File "/var/venv/radar/local/lib/python2.7/site-packages/kombu/serialization.py", line 59, in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
  File "/var/venv/radar/local/lib/python2.7/site-packages/kombu/serialization.py", line 55, in _reraise_errors
    yield
  File "/var/venv/radar/local/lib/python2.7/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/var/venv/radar/local/lib/python2.7/site-packages/msgpack/__init__.py", line 47, in packb
    return Packer(**kwargs).pack(o)
  File "_packer.pyx", line 224, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:224)
  File "_packer.pyx", line 226, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:226)
  File "_packer.pyx", line 214, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:214)
  File "_packer.pyx", line 221, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:221)
EncodeError: can't serialize <ExceptionInfo: UnicodeDecodeError('ascii', 'Article non trouv\xc3\xa9', 17, 18, 'ordinal not in range(128)')>
```

I'm not sure if the two errors are related, but at least in the first one it looks like Celery tries to read msgpack encoded binary content from Redis while thinking it's a string.

Unfortunately so far I could not reproduce the issue, but it occured several times, even after worker restarts. I'll upgrade to 3.1.17, but didn't find anything related in the changelogs. Maybe someone knows what's going on here? Could it be related to `new_join=1`?

**Comment by dbrgn** [comment]
I started debugging this issue again and found the configuration change that causes the problem.

This is the example I used for testing:

```python
# -*- coding: utf-8 -*-
from celery import Celery

app = Celery('tasks', broker='redis://localhost:6379/0')
app.conf.CELERY_TASK_SERIALIZER = 'json'
app.conf.CELERY_RESULT_SERIALIZER = 'pickle'
app.conf.CELERY_ACCEPT_CONTENT = ['json', 'pickle']
app.conf.CELERY_RESULT_BACKEND = 'redis://localhost:6379/0?new_join=1'

@app.task
def add(x, y):
    summe = x + y
    return '%s + %s = %s' % (x, y, summe)

@app.task
def err():
    return 'füübär'.encode('ascii')

@app.task
def combine(a, b):
    return '%s | %s' % (a, b)
```

With this configuration, when calling a chord (`chord([add.s(1, 2), err.s()])(combine.s())`), I get a proper exception trace in the worker log:

```
MainProcess] celery@adalbert ready.
[2015-04-22 15:17:16,272: INFO/MainProcess] Received task: cel.add[388e55ca-940b-4a25-944b-4f6c724257bb]
[2015-04-22 15:17:16,275: INFO/MainProcess] Received task: cel.err[2f4a1eb6-f279-43ee-a758-fb9dcb03d327]
[2015-04-22 15:17:16,284: INFO/MainProcess] Task cel.add[388e55ca-940b-4a25-944b-4f6c724257bb] succeeded in 0.008015557075850666s: '1 + 2 = 3'
[2015-04-22 15:17:16,287: ERROR/Worker-2] Chord callback for '5bf4ca43-a932-4c67-bbf9-682aa27ff5d6' raised: ChordError("Dependency 2f4a1eb6-f279-43ee-a758-fb9dcb03d327 raised <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')>",)
Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/redis.py", line 224, in _new_chord_return
    callback.delay([unpack(tup, decode) for tup in resl])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/redis.py", line 224, in <listcomp>
    callback.delay([unpack(tup, decode) for tup in resl])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/redis.py", line 187, in _unpack_chord_result
    raise ChordError('Dependency {0} raised {1!r}'.format(tid, retval))
celery.exceptions.ChordError: Dependency 2f4a1eb6-f279-43ee-a758-fb9dcb03d327 raised <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')>
[2015-04-22 15:17:16,290: ERROR/MainProcess] Task cel.err[2f4a1eb6-f279-43ee-a758-fb9dcb03d327] raised unexpected: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')
Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/tmp/cel.py", line 17, in err
    return 'füübär'.encode('ascii')
UnicodeEncodeError: 'ascii' codec can't encode characters in position 1-2: ordinal not in range(128)
```

But when changing the `CELERY_RESULT_SERIALIZER` to `json`, serialization fails:

```
[2015-04-22 15:18:13,024: WARNING/MainProcess] celery@adalbert ready.
[2015-04-22 15:18:15,069: INFO/MainProcess] Received task: cel.add[3eda63c6-f027-4f69-b367-0305d448000d]
[2015-04-22 15:18:15,071: INFO/MainProcess] Received task: cel.err[60758a42-8392-4092-b950-b133b64ddacc]
[2015-04-22 15:18:15,080: INFO/MainProcess] Task cel.add[3eda63c6-f027-4f69-b367-0305d448000d] succeeded in 0.007941085961647332s: '1 + 2 = 3'
[2015-04-22 15:18:15,083: WARNING/Worker-2] /home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/app/trace.py:365: RuntimeWarning: Exception raised outside body: EncodeError(TypeError(TypeError("<ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable",),),):
Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 87, in dumps
    return self._encode(data)
  File "/usr/lib64/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 55, in _reraise_errors
    yield
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 141, in dumps
    return implementation.dumps(value)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 89, in dumps
    raise TypeError(TypeError(*exc.args)).with_traceback(sys.exc_info()[2])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 87, in dumps
    return self._encode(data)
  File "/usr/lib64/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/app/trace.py", line 296, in trace_task
    on_chord_part_return(task, state, R)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/redis.py", line 209, in _new_chord_return
    .rpush(jkey, self.encode([1, tid, state, result]))          \
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/base.py", line 182, in encode
    _, _, payload = dumps(data, serializer=self.serializer)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/usr/lib64/python3.4/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 59, in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/five.py", line 132, in reraise
    raise value.with_traceback(tb)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 55, in _reraise_errors
    yield
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 141, in dumps
    return implementation.dumps(value)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 89, in dumps
    raise TypeError(TypeError(*exc.args)).with_traceback(sys.exc_info()[2])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 87, in dumps
    return self._encode(data)
  File "/usr/lib64/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
kombu.exceptions.EncodeError: <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable

  exc, exc_info.traceback)))
[2015-04-22 15:18:15,085: CRITICAL/MainProcess] Task cel.err[60758a42-8392-4092-b950-b133b64ddacc] INTERNAL ERROR: EncodeError(TypeError(TypeError("<ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable",),),)
Traceback (most recent call last):
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/app/trace.py", line 296, in trace_task
    on_chord_part_return(task, state, R)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/redis.py", line 209, in _new_chord_return
    .rpush(jkey, self.encode([1, tid, state, result]))          \
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/celery/backends/base.py", line 182, in encode
    _, _, payload = dumps(data, serializer=self.serializer)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/usr/lib64/python3.4/contextlib.py", line 77, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 59, in _reraise_errors
    reraise(wrapper, wrapper(exc), sys.exc_info()[2])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/five.py", line 132, in reraise
    raise value.with_traceback(tb)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 55, in _reraise_errors
    yield
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/kombu/serialization.py", line 164, in dumps
    payload = encoder(data)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 141, in dumps
    return implementation.dumps(value)
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 89, in dumps
    raise TypeError(TypeError(*exc.args)).with_traceback(sys.exc_info()[2])
  File "/home/danilo/.virtualenvs/test/lib/python3.4/site-packages/anyjson/__init__.py", line 87, in dumps
    return self._encode(data)
  File "/usr/lib64/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib64/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib64/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib64/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
kombu.exceptions.EncodeError: <ExceptionInfo: UnicodeEncodeError('ascii', 'füübär', 1, 3, 'ordinal not in range(128)')> is not JSON serializable
```

As far as I could tell, this happens with a chord, but not when calling a regular task directly.

**Comment by somada141** [comment]
Ok I might be entirely off here but I think its worth mentioning. I encountered a very similar problem yday at work where I was getting an error very similar to:

```
UnicodeEncodeError('ascii', '<Some UTF8 stuff>', 1, 3, 'ordinal not in range(128)')
```

I should mention that I was using Celery 3.1.18 with a Redis backend and a RabbitMQ broker.

After hours of trying to figure the issue out with my colleague he realised that the error was an exception thrown within an exception's `except` clause rather than a first-class exception. The issue was coming from the Redis backend running out of memory and trying to report on the fact by printing the Celery task result alongside an exception message. 

The task result however contained UTF8 encoded strings and [the `annotate_exception` method used in `redis-py` 2.1.3](https://github.com/andymccurdy/redis-py/blob/4d0b0afe9c9a431ed50c3e9fb95a0aa88b1f1038/redis/client.py#L2529) tries to force it to unicode causing a subsequent `UnicodeEncodeError` to be raised within the `except` clause. As a result of the exception-within-exception, the original Redis error was lost to us. 

The above bug has been fixed on `redis-py` 2.1.5 and the issue was resolved upon upgrading the package. In my case it was technically not a Celery issue but a known `redis-py` issue.

While your issue may not be coming from Redis running out of memory (not everyone's as careless as I was) it might well be related to another Redis error that gets obscured by UTF8/Unicode rigamarole. 

Hope this saves someone the 5 hours it cost us :)
